banyan/src/forest/read.rs

```diff
@@ -19,7 +13,295 @@ use anyhow::{anyhow, Result};
 use core::fmt::Debug;
 use futures::{prelude::*, stream::BoxStream};
 use libipld::cbor::DagCbor;
+use smallvec::{smallvec, SmallVec};
+
 use std::{iter, sync::Arc, time::Instant};
+enum Mode {
+    Forward,
+    Backward,
+}
+pub(crate) struct ForestIter<T: TreeTypes, V, R, Q: Query<T>, F> {
+    forest: Forest<T, V, R>,
+    offset: u64,
+    query: Q,
+    mk_extra: F,
+    stack: SmallVec<[TraverseState<T>; 5]>,
+    mode: Mode,
+}
+
+struct TraverseState<T: TreeTypes> {
+    index: Arc<Index<T>>,
+    // If `index` points to a branch node, `position` points to the currently
+    // traversed child
+    position: usize,
```

Comment:
It is probably worth noting that this is not the child offset, but the child offset in case of forward traversal, and the child offset + 1 in case of backward traversal.

I guess you have already tried to make this an `isize` and have it be exactly the position?

Reply:
I tried this out here: https://github.com/Actyx/banyan/commit/10bfa3dd011f208c62d2ebd5a5e85a1802638ce8

Reply:
Cool! Let's get it in!
I haven't tried that. At some point I kind of lost sight ..